src/internal/pfsdb/commits.go

```diff
@@ -0,0 +1,956 @@
+type ErrCommitNotFound struct {
+	RowID    CommitID
+	CommitID string
+}
+
+// Error satisfies the error interface.
+func (err ErrCommitNotFound) Error() string {
+	if id := err.RowID; id != 0 {
+		return fmt.Sprintf("commit id=%d not found", id)
+	}
+	return fmt.Sprintf("commit %s not found", err.CommitID)
+}
+
+func (err ErrCommitNotFound) GRPCStatus() *status.Status {
+	return status.New(codes.NotFound, err.Error())
+}
+
+// ErrParentCommitNotFound is returned when a commit's parent is not found in postgres.
+type ErrParentCommitNotFound struct {
+	Repo          string
+	ChildRowID    CommitID
+	ChildCommitID string
+}
+
+func IsParentCommitNotFound(err error) bool {
+	return dbutil.IsNotNullViolation(err, "from_id")
+}
+
+// Error satisfies the error interface.
+func (err ErrParentCommitNotFound) Error() string {
+	if id := err.ChildRowID; id != 0 {
+		return fmt.Sprintf("parent commit of commit id=%d not found", id)
+	}
+	return fmt.Sprintf("parent commit of commit %s/%s not found", err.Repo, err.ChildCommitID)
+}
+
+func (err ErrParentCommitNotFound) GRPCStatus() *status.Status {
+	return status.New(codes.NotFound, err.Error())
+}
+
+// ErrChildCommitNotFound is returned when a commit's child is not found in postgres.
+type ErrChildCommitNotFound struct {
+	Repo           string
+	ParentRowID    CommitID
+	ParentCommitID string
+}
+
+func IsChildCommitNotFound(err error) bool {
+	return dbutil.IsNotNullViolation(err, "to_id")
+}
+
+// Error satisfies the error interface.
+func (err ErrChildCommitNotFound) Error() string {
+	if id := err.ParentRowID; id != 0 {
+		return fmt.Sprintf("parent commit of commit id=%d not found", id)
+	}
+	return fmt.Sprintf("parent commit of commit %s/%s not found", err.Repo, err.ParentCommitID)
+}
+
+func (err ErrChildCommitNotFound) GRPCStatus() *status.Status {
+	return status.New(codes.NotFound, err.Error())
+}
+
+func IsErrCommitNotFound(err error) bool {
+	return errors.As(err, &ErrCommitNotFound{})
+}
+
+// ErrCommitMissingInfo is returned by CreateCommit() when a commitInfo is missing a field.
+type ErrCommitMissingInfo struct {
+	Field string
+}
+
+func (err ErrCommitMissingInfo) Error() string {
+	return fmt.Sprintf("commitInfo.%s is missing/nil", err.Field)
+}
+
+func (err ErrCommitMissingInfo) GRPCStatus() *status.Status {
+	return status.New(codes.FailedPrecondition, err.Error())
+}
+
+// ErrCommitAlreadyExists is returned by CreateCommit() when a commit with the same name already exists in postgres.
+type ErrCommitAlreadyExists struct {
+	CommitID string
+}
+
+// Error satisfies the error interface.
+func (err ErrCommitAlreadyExists) Error() string {
+	if id := err.CommitID; id != "" {
+		return fmt.Sprintf("commit %s already exists", id)
+	}
+	return "commit already exists"
+}
+
+func (err ErrCommitAlreadyExists) GRPCStatus() *status.Status {
+	return status.New(codes.AlreadyExists, err.Error())
+}
+
+type commitRow struct {
```

Comment:
Update `Commit` in `model.go` and use that instead.

Reply:
I think there's value in having a type that exactly matches a row in the commits table.  `commitRow` is a good name for that.

This struct seems to include fields which are not in the commits table though.
- `child_commits` and `parent_commits`
- `repo_name`
- `branch_name`

Reply:
Okay after reading around more, it looks like this object is used for reading the results of a join.  I agree with albert that the `Commit` object is a better place for that.

Maybe we could have the Commit object embed the commitRow?

```
type Commit struct {
   commitRow
  // other fields.
}
```

Reply:
Yes, the whole point of having `models.go` is to define structs that make it easier to deserialize data from the database. An added benefit is that the structs embedded each other like `Repo` embeds `Project`, which clearly captures our data model in code.